### PR TITLE
Read IO plugin configuration (openPMD plugin) from reading applications (TOML)

### LINKED ISSUE #3717: Read IO plugin configuration (openPMD plugin) from reading applications (TOML)
**Status Quo:** The openPMD plugin is configured via command-line parameters of PIConGPU, including the period of activation and the written data.

**Goal:** Reading applications (data sinks, analyses, visualizations, post-processing) state their data requirements at the simulation's start (in which frequency is which data needed). PIConGPU collects these data requirements, merges them into one instance of the openPMD plugin. The openPMD plugin produces data only if a data source (species_all, e_all, fields_all, E, B, ...) has been requested by at least one reader for the current step. Unlike currently, written data sources may vary across iterations in the same instance of the openPMD plugin.

**Suggested schema:**
The openPMD plugin currently has the following command line parameters:
```
--openPMD.period                   // per data sink
--openPMD.source                   // per data sink
--openPMD.compression              // deprecated, use JSON
--openPMD.file                     // per instance of the plugin
--openPMD.ext                      // per instance of the plugin
--openPMD.infix                    // per instance of the plugin
--openPMD.json                     // per instance of the plugin
--openPMD.dataPreparationStrategy  // per instance of the plugin
```
As noted above, some of those parameters make sense to specify per data sink, while others should only be given once per instance of the openPMD plugin to avoid conflicts (specifying JSON twice.. what should we do? merge?, what if some readers specify a JSON file? the `--openPMD.file` must be the same for one instance of the plugin).

I propose the following configuration:
**PIConGPU**
```
--openPMD.period dynamic // maybe leave this one out
--openPMD.name simData
--openPMD.json @path/to/config.json
--openPMD.toml "path/to/sink1.toml;path/to/sink2.toml"

# second instance of the openPMD plugin if people like that
--openPMD.period dynamic
--openPMD.name whyAreYouDoingThis
--openPMD.toml "i/want/to/have/an/instance/for/myself.toml"
```

**Sink 1** would write the following content to `path/to/sink1.toml`:
```toml
[period]
200 = ["E", "B"]
400 = ["fields_all"]
```

**Sink 2** would write the following content to `path/to/sink2.toml`:
```toml
[period]
100 = ["e_all", "i_all"]
300 = ["species_all"]
```

**Sink 3** would write the following content to `i/want/to/have/an/instance/for/myself.toml`:
```toml
[period]
1 = ["species_all", "fields_all"] # sink 3 is a greedy one
```

**Explanation**
A sink is linked to an instance of the openPMD plugin via the TOML file:

* The sink writes the TOML file
* The openPMD plugin is told about all of the TOML files that it should read via a command line parameter

No configuration is duplicated.

Upon startup, the openPMD plugin will wait until all TOML files are present in the file system and proceed to read its data requirements from there. All other configuration is the same as it is now. For each data requirement `requirement=(period, sources)`, the openPMD plugin will output all of the configured `sources` in each step divisible by `period`. If multiple requirements apply for a single simulation step, the data is merged into a single output.

### PR DESCRIPTION
**Pull request by franzpoeschel**
Close #3717, see there for a detailed description

TODO:

- [x] Add thirdparty library and integrate it with the build system
- [x] Implement new logic in openPMD plugin
- [x] Actually read config from TOML files (currently, a little example is hardcoded)
- [x] Check that `selfRegister==true` is handled correctly (checkpoint writing)
    keep the `.toml` option out of `--checkpoint.openPMD` somehow?
- [x] Discuss command line parameter schema (may drop the `--openPMD.period dynamic` thing, it's easy to implement for now) Update: `--openPMD.period` and `--openPMD.toml` are now mutually exclusive. If using TOML, then the parameter must be the only parameter.
- [x] Logging
- [x] parallel reading
- [x] Timeout for waiting on files?
- [x] Add documentation
- [x] Merge #3906 first
- [x] I'd appreciate if someone could test that your multiplugin workflows still work
- [x] rebase after #3919 to avoid documentation merge conflicts

### CODE REVIEW

File: include/picongpu/plugins/openPMD/toml.cpp [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+        {
+            std::set<std::string> result;
+            for(uint32_t period : m_nextActiveAt.begin()->second)
+            {
+                for(std::string const& source : m_period.at(period))
+                {
+                    result.insert(source);
+                }
+            }
+            return {result.end(), result.end()};
```
Comment (franzpoeschel) [inline include/picongpu/plugins/openPMD/toml.cpp:52]: Note to self: result.begin()

**Comment by franzpoeschel** [comment]
Just did a little test run with this thing:

First TOML file:
```toml
[period]
200 = ["E", "B"]
400 = "fields_all"
```

Second one:
```toml
[period]
500 = "species_all"
```
Resulting dataset:
```
  float     /data/0/fields/B/x                                        {32, 32, 32}
  float     /data/0/fields/B/y                                        {32, 32, 32}
  float     /data/0/fields/B/z                                        {32, 32, 32}
  float     /data/0/fields/E/x                                        {32, 32, 32}
  float     /data/0/fields/E/y                                        {32, 32, 32}
  float     /data/0/fields/E/z                                        {32, 32, 32}
  float     /data/0/fields/e_all_chargeDensity                        {32, 32, 32}
  float     /data/0/fields/e_all_energyDensity                        {32, 32, 32}
  float     /data/0/fields/e_all_particleMomentumComponent            {32, 32, 32}
  float     /data/0/fields/i_all_chargeDensity                        {32, 32, 32}
  float     /data/0/fields/i_all_energyDensity                        {32, 32, 32}
  float     /data/0/fields/i_all_particleMomentumComponent            {32, 32, 32}
  uint64_t  /data/0/fields/picongpu_idProvider/nextId                 {1, 1, 1}
  uint64_t  /data/0/fields/picongpu_idProvider/startId                {1, 1, 1}
  float     /data/0/particles/e/momentum/x                            {819200}
  float     /data/0/particles/e/momentum/y                            {819200}
  float     /data/0/particles/e/momentum/z                            {819200}
  uint64_t  /data/0/particles/e/particlePatches/extent/x              {1}
  uint64_t  /data/0/particles/e/particlePatches/extent/y              {1}
  uint64_t  /data/0/particles/e/particlePatches/extent/z              {1}
  uint64_t  /data/0/particles/e/particlePatches/numParticles          {1}
  uint64_t  /data/0/particles/e/particlePatches/numParticlesOffset    {1}
  uint64_t  /data/0/particles/e/particlePatches/offset/x              {1}
  uint64_t  /data/0/particles/e/particlePatches/offset/y              {1}
  uint64_t  /data/0/particles/e/particlePatches/offset/z              {1}
  float     /data/0/particles/e/position/x                            {819200}
  float     /data/0/particles/e/position/y                            {819200}
  float     /data/0/particles/e/position/z                            {819200}
  int32_t   /data/0/particles/e/positionOffset/x                      {819200}
  int32_t   /data/0/particles/e/positionOffset/y                      {819200}
  int32_t   /data/0/particles/e/positionOffset/z                      {819200}
  float     /data/0/particles/e/weighting                             {819200}
  float     /data/0/particles/i/momentum/x                            {819200}
  float     /data/0/particles/i/momentum/y                            {819200}
  float     /data/0/particles/i/momentum/z                            {819200}
  uint64_t  /data/0/particles/i/particlePatches/extent/x              {1}
  uint64_t  /data/0/particles/i/particlePatches/extent/y              {1}
  uint64_t  /data/0/particles/i/particlePatches/extent/z              {1}
  uint64_t  /data/0/particles/i/particlePatches/numParticles          {1}
  uint64_t  /data/0/particles/i/particlePatches/numParticlesOffset    {1}
  uint64_t  /data/0/particles/i/particlePatches/offset/x              {1}
  uint64_t  /data/0/particles/i/particlePatches/offset/y              {1}
  uint64_t  /data/0/particles/i/particlePatches/offset/z              {1}
  float     /data/0/particles/i/position/x                            {819200}
  float     /data/0/particles/i/position/y                            {819200}
  float     /data/0/particles/i/position/z                            {819200}
  int32_t   /data/0/particles/i/positionOffset/x                      {819200}
  int32_t   /data/0/particles/i/positionOffset/y                      {819200}
  int32_t   /data/0/particles/i/positionOffset/z                      {819200}
  float     /data/0/particles/i/weighting                             {819200}
  float     /data/200/fields/B/x                                      {32, 32, 32}
  float     /data/200/fields/B/y                                      {32, 32, 32}
  float     /data/200/fields/B/z                                      {32, 32, 32}
  float     /data/200/fields/E/x                                      {32, 32, 32}
  float     /data/200/fields/E/y                                      {32, 32, 32}
  float     /data/200/fields/E/z                                      {32, 32, 32}
  uint64_t  /data/200/fields/picongpu_idProvider/nextId               {1, 1, 1}
  uint64_t  /data/200/fields/picongpu_idProvider/startId              {1, 1, 1}
  float     /data/400/fields/B/x                                      {32, 32, 32}
  float     /data/400/fields/B/y                                      {32, 32, 32}
  float     /data/400/fields/B/z                                      {32, 32, 32}
  float     /data/400/fields/E/x                                      {32, 32, 32}
  float     /data/400/fields/E/y                                      {32, 32, 32}
  float     /data/400/fields/E/z                                      {32, 32, 32}
  float     /data/400/fields/e_all_chargeDensity                      {32, 32, 32}
  float     /data/400/fields/e_all_energyDensity                      {32, 32, 32}
  float     /data/400/fields/e_all_particleMomentumComponent          {32, 32, 32}
  float     /data/400/fields/i_all_chargeDensity                      {32, 32, 32}
  float     /data/400/fields/i_all_energyDensity                      {32, 32, 32}
  float     /data/400/fields/i_all_particleMomentumComponent          {32, 32, 32}
  uint64_t  /data/400/fields/picongpu_idProvider/nextId               {1, 1, 1}
  uint64_t  /data/400/fields/picongpu_idProvider/startId              {1, 1, 1}
  uint64_t  /data/500/fields/picongpu_idProvider/nextId               {1, 1, 1}
  uint64_t  /data/500/fields/picongpu_idProvider/startId              {1, 1, 1}
  uint64_t  /data/500/particles/e/particlePatches/extent/x            {1}
  uint64_t  /data/500/particles/e/particlePatches/extent/y            {1}
  uint64_t  /data/500/particles/e/particlePatches/extent/z            {1}
  uint64_t  /data/500/particles/e/particlePatches/numParticles        {1}
  uint64_t  /data/500/particles/e/particlePatches/numParticlesOffset  {1}
  uint64_t  /data/500/particles/e/particlePatches/offset/x            {1}
  uint64_t  /data/500/particles/e/particlePatches/offset/y            {1}
  uint64_t  /data/500/particles/e/particlePatches/offset/z            {1}
  uint64_t  /data/500/particles/i/particlePatches/extent/x            {1}
  uint64_t  /data/500/particles/i/particlePatches/extent/y            {1}
  uint64_t  /data/500/particles/i/particlePatches/extent/z            {1}
  uint64_t  /data/500/particles/i/particlePatches/numParticles        {1}
  uint64_t  /data/500/particles/i/particlePatches/numParticlesOffset  {1}
  uint64_t  /data/500/particles/i/particlePatches/offset/x            {1}
  uint64_t  /data/500/particles/i/particlePatches/offset/y            {1}
  uint64_t  /data/500/particles/i/particlePatches/offset/z            {1}
  float     /data/600/fields/B/x                                      {32, 32, 32}
  float     /data/600/fields/B/y                                      {32, 32, 32}
  float     /data/600/fields/B/z                                      {32, 32, 32}
  float     /data/600/fields/E/x                                      {32, 32, 32}
  float     /data/600/fields/E/y                                      {32, 32, 32}
  float     /data/600/fields/E/z                                      {32, 32, 32}
  uint64_t  /data/600/fields/picongpu_idProvider/nextId               {1, 1, 1}
  uint64_t  /data/600/fields/picongpu_idProvider/startId              {1, 1, 1}
```

File: include/picongpu/plugins/openPMD/toml.cpp [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+                {
+                    auto dataSource
+                        = toml::expect<std::string>(value)
+                              .or_else(
+                                  [](auto const&) -> toml::success<std::string>
+                                  {
+                                      throw std::runtime_error("[openPMD plugin] Data sources in TOML "
+                                                               "file must be a string or a vector of strings.");
+                                  })
+                              .value;
```
Comment (franzpoeschel) [inline include/picongpu/plugins/openPMD/toml.cpp:127]: remove leftover stuff

File: include/picongpu/plugins/openPMD/toml.cpp [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+    }
+} // namespace
+
+
+namespace picongpu
+{
+    namespace toml
+    {
+        using namespace std::literals::chrono_literals;
+        DataSources::DataSources(std::string tomlFiles) : m_period{waitForParseAndMergeTomlFiles(tomlFiles, 5s)}
```
Comment (franzpoeschel) [inline include/picongpu/plugins/openPMD/toml.cpp:200]: unneeded copy

File: include/picongpu/plugins/openPMD/toml.cpp (hunk 2 above)
Comment (franzpoeschel) [reply include/picongpu/plugins/openPMD/toml.cpp:127]: eh, no removing
that one stays

**Comment by franzpoeschel** [comment]
The latest commit fixes waiting for late files and adds logging for that:
```
PIConGPUVerbose INPUT_OUTPUT(32) | openPMD: Reading data requirements from TOML files:                                                                                               
        [/home/franzpoeschel/src/singularity/toml1.toml, /home/franzpoeschel/src/singularity/toml0.toml]                                                                            
PIConGPUVerbose INPUT_OUTPUT(32) | openPMD: Still waiting for TOML files:                                                                                                            
        [/home/franzpoeschel/src/singularity/toml0.toml]                                                                                                                            
PIConGPUVerbose INPUT_OUTPUT(32) | openPMD: Still waiting for TOML files:
        [/home/franzpoeschel/src/singularity/toml0.toml]
PIConGPUVerbose INPUT_OUTPUT(32) | openPMD: Still waiting for TOML files:
        [/home/franzpoeschel/src/singularity/toml0.toml]
PIConGPUVerbose INPUT_OUTPUT(32) | openPMD: opening Series openPMD/stream
PIConGPUVerbose INPUT_OUTPUT(32) | openPMD: open file: openPMD/stream.bp
```

Next up: parallel reading (i.e. non-parallel reading)

**Comment by franzpoeschel** [comment]
(PR stalled for now until René is back from holiday to review this approach.)

**Comment by franzpoeschel** [comment]
Discussed offline with René:
We will change this such that one instance of the openPMD plugin can be configured entirely by one single TOML file. This file will have sections for the single reading codes' data requirements:
```toml
[openPMD]
"file" = simOutput
"ext" = "bp"
"backendConfig" = "@./adios2_config.json"

[sink.analysis_code_name.period]
200 = ["E", "B"]
400 = "fields_all"

[sink.other_analysis_code_name.period]
500 = "species_all"
```

Under the `openPMD` group, we simply put all classical cmd line parameters of the openPMD plugin. Under `[sink.<arbitrary name>.period]`, the single reading codes' data requirements are found. This allows us to add other sink-specific keys in future:
```toml
# this will not be part of this PR
# just demonstrating that this schema is extensible
[sink.analysis_code_name]
"logging" = true
```

Other point discussed: Allow time slicing syntax

**Comment by franzpoeschel** [comment]
Alright, command line parameter handling should now work. Still need to test that checkpoint/restarting is handled correctly, and I might need some help so that this PR does not also add a parameter `--checkpoint.openPMD.toml`. The plugin checks that `toml` is the only parameter specifies and fails if any other parameter has been passed.

**Comment by PrometheusPi** [comment]
@franzpoeschel What is the status of this pul request?

**Comment by franzpoeschel** [comment]
@PrometheusPi I meet with René next week to discuss the current status. If we agree that the approach taken by this is ok, I expect it can be merged soon.

Note that this PR is **not** about TOML configuration for the openPMD-api, but for PIConGPU's openPMD plugin.

**Comment by franzpoeschel** [comment]
I added the things we talked about yesterday @psychocoderHPC:
- [ ] Allow to call the openPMD plugin with `--openPMD.toml "" --openPMD.file output --openPMD.period 100`, e.g. consider empty strings as equivalent to no passed parameters
  I'd generally appreciate if someone who uses the openPMD plugin as a multi plugin could check if things are still running.
- [ ] Added documentation. I've documented most of it with inline comments inside a [sample `.toml` file](https://github.com/ComputationalRadiationPhysics/picongpu/blob/2a358739be22d271e41c2d681fa55eceae0e4a86/docs/source/usage/plugins/openPMD.toml)
- [ ] Make the `toml::DataSources` stateless

**Comment by psychocoderHPC** [comment]
I tested that multi-plugin usage with command line parameters still works.
We need to rebase this PR after #3919 is merged because of conflicts in the documentation.

File: include/picongpu/plugins/common/MPIHelpers.hpp [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with PIConGPU.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include <vector>
```
Comment (psychocoderHPC) [inline include/picongpu/plugins/common/MPIHelpers.hpp:22]: `#include <string>` is missing

**Review by psychocoderHPC — CHANGES_REQUESTED** [review]

**Comment by psychocoderHPC** [comment]
PR can be rebased

**Comment by franzpoeschel** [comment]
> PR can be rebased

done

**Comment by psychocoderHPC** [comment]
@franzpoeschel What do you mean by `Timeout for waiting on files?`

**Comment by franzpoeschel** [comment]
> @franzpoeschel What do you mean by `Timeout for waiting on files?`

In our intended workflows, the TOML file is created asynchronously by the reading codes. So, as long as the file is not there yet, [these lines here](https://github.com/ComputationalRadiationPhysics/picongpu/blob/c4a367f7222ad658fafa854e78eb332862fdb559/include/picongpu/plugins/openPMD/toml.cpp#L232) will sleep until the file is found. Should I keep that part? If yes, I think I should add a timeout. @psychocoderHPC

**Comment by psychocoderHPC** [comment]
> > @franzpoeschel What do you mean by `Timeout for waiting on files?`
> 
> In our intended workflows, the TOML file is created asynchronously by the reading codes. So, as long as the file is not there yet, [these lines here](https://github.com/ComputationalRadiationPhysics/picongpu/blob/c4a367f7222ad658fafa854e78eb332862fdb559/include/picongpu/plugins/openPMD/toml.cpp#L232) will sleep until the file is found. Should I keep that part? If yes, I think I should add a timeout. @psychocoderHPC

Ohh yes, a timeout makes sense. IMO set as default 5min.

**Comment by franzpoeschel** [comment]
> > > @franzpoeschel What do you mean by `Timeout for waiting on files?`
> > 
> > 
> > In our intended workflows, the TOML file is created asynchronously by the reading codes. So, as long as the file is not there yet, [these lines here](https://github.com/ComputationalRadiationPhysics/picongpu/blob/c4a367f7222ad658fafa854e78eb332862fdb559/include/picongpu/plugins/openPMD/toml.cpp#L232) will sleep until the file is found. Should I keep that part? If yes, I think I should add a timeout. @psychocoderHPC
> 
> Ohh yes, a timeout makes sense. IMO set as default 5min.

Done and tested

**Review by steindev — APPROVED** [review]

**Pull request merged**